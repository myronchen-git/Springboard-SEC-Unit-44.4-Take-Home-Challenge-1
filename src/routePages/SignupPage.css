.Signup {
  background-color: white;

  margin: 2em;
  padding: 2em;

  border-radius: 2em;
  box-shadow: 0 0 0.75em grey;
}

.Signup__form {
  display: flex;
  flex-direction: column;
}

.Signup button {
  margin-bottom: 1em;
}
