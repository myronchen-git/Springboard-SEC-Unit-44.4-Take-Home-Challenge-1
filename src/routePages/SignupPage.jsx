import { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Alert, Button, Form, FormGroup, Input, Label } from 'reactstrap';

import { UserContext } from '../contexts.jsx';

// ==================================================

/**
 * Displays and handles the user registration form.
 */
function SignupPage() {
  const initialFormData = {
    firstName: '',
    lastName: '',
    email: '',
  };
  const [formData, setFormData] = useState(initialFormData);
  const [errorMessage, setErrorMessage] = useState(null);
  const { signup } = useContext(UserContext);
  const navigate = useNavigate();

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();

    try {
      await signup(formData);
    } catch (err) {
      setErrorMessage(err.message);
      return;
    }

    navigate('/');
  }

  return (
    <main className="Signup">
      <Form className="Signup__form" onSubmit={handleSubmit}>
        <FormGroup>
          <Label htmlFor="Signup__input-first-name">First Name</Label>
          <Input
            id="Signup__input-first-name"
            type="text"
            name="firstName"
            value={formData.firstName}
            required
            onChange={handleChange}
          ></Input>
        </FormGroup>
        <FormGroup>
          <Label htmlFor="Signup__input-last-name">Last Name</Label>
          <Input
            id="Signup__input-last-name"
            type="text"
            name="lastName"
            value={formData.lastName}
            required
            onChange={handleChange}
          ></Input>
        </FormGroup>
        <FormGroup>
          <Label htmlFor="Signup__input-email">Email</Label>
          <Input
            id="Signup__input-email"
            type="email"
            name="email"
            value={formData.email}
            required
            onChange={handleChange}
          ></Input>
        </FormGroup>
        <Button type="submit">Submit</Button>
        {errorMessage && <Alert color="danger">{errorMessage}</Alert>}
      </Form>
    </main>
  );
}

// ==================================================

export default SignupPage;
