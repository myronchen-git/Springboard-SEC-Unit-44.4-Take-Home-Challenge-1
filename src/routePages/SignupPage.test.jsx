import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useNavigate } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { UserContext } from '../contexts.jsx';
import SignupPage from './SignupPage.jsx';

// ==================================================

vi.mock('react-router-dom');

// ==================================================

describe('SignupPage', () => {
  const newData = Object.freeze({
    firstName: 'First',
    lastName: 'Last',
    email: 'email@email.com',
  });

  const mockSignup = vi.fn();
  const mockNavigate = vi.fn();
  useNavigate.mockReturnValue(mockNavigate);

  beforeEach(() => {
    mockSignup.mockReset();
    mockNavigate.mockReset();
  });

  it('renders.', () => {
    render(
      <UserContext.Provider value={{ signup: mockSignup }}>
        <SignupPage />
      </UserContext.Provider>
    );
  });

  it('matches snapshot.', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ signup: mockSignup }}>
        <SignupPage />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('can update form data.', async () => {
    // Arrange
    const user = userEvent.setup();

    const { getByLabelText } = render(
      <UserContext.Provider value={{ signup: mockSignup }}>
        <SignupPage />
      </UserContext.Provider>
    );

    const firstNameInput = getByLabelText('First Name');
    const lastNameInput = getByLabelText('Last Name');
    const emailInput = getByLabelText('Email');

    // Act
    await user.click(firstNameInput);
    await user.keyboard(newData.firstName);
    await user.click(lastNameInput);
    await user.keyboard(newData.lastName);
    await user.click(emailInput);
    await user.keyboard(newData.email);

    // Assert
    expect(firstNameInput).toHaveValue(newData.firstName);
    expect(lastNameInput).toHaveValue(newData.lastName);
    expect(emailInput).toHaveValue(newData.email);

    expect(mockSignup).not.toHaveBeenCalled();
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it('can submit form data.', async () => {
    // Arrange
    const user = userEvent.setup();

    const { getByLabelText, getByText } = render(
      <UserContext.Provider value={{ signup: mockSignup }}>
        <SignupPage />
      </UserContext.Provider>
    );

    await user.click(getByLabelText('First Name'));
    await user.keyboard(newData.firstName);
    await user.click(getByLabelText('Last Name'));
    await user.keyboard(newData.lastName);
    await user.click(getByLabelText('Email'));
    await user.keyboard(newData.email);

    const submitBtn = getByText('Submit');

    // Act
    await user.click(submitBtn);

    // Assert
    expect(mockSignup).toHaveBeenCalledOnce();
    expect(mockSignup).toHaveBeenCalledWith(newData);
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });

  it('displays an error message if there is an issue when signing up.', async () => {
    // Arrange
    const errorMessage = 'error';
    mockSignup.mockRejectedValue({ message: errorMessage });

    const user = userEvent.setup();

    const { getByLabelText, getByText } = render(
      <UserContext.Provider value={{ signup: mockSignup }}>
        <SignupPage />
      </UserContext.Provider>
    );

    await user.click(getByLabelText('First Name'));
    await user.keyboard(newData.firstName);
    await user.click(getByLabelText('Last Name'));
    await user.keyboard(newData.lastName);
    await user.click(getByLabelText('Email'));
    await user.keyboard(newData.email);

    const submitBtn = getByText('Submit');

    // Act
    await user.click(submitBtn);

    // Assert
    expect(getByText(errorMessage)).toBeVisible();

    expect(mockSignup).toHaveBeenCalledOnce();
    expect(mockSignup).toHaveBeenCalledWith(newData);
    expect(mockNavigate).not.toHaveBeenCalled();
  });
});
