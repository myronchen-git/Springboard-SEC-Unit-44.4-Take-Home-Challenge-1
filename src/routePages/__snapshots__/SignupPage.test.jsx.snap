// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignupPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="Signup"
  >
    <title>
      Yodlr Registration Portal
    </title>
    <h1>
      Yodlr Registration Portal
    </h1>
    bound HTMLFormElement {
      "0": <input
        class="form-control"
        id="Signup__input-first-name"
        name="firstName"
        required=""
        type="text"
        value=""
      />,
      "1": <input
        class="form-control"
        id="Signup__input-last-name"
        name="lastName"
        required=""
        type="text"
        value=""
      />,
      "2": <input
        class="form-control"
        id="Signup__input-email"
        name="email"
        required=""
        type="email"
        value=""
      />,
      "3": <button
        class="btn btn-secondary"
        type="submit"
      >
        Submit
      </button>,
    }
  </main>
</DocumentFragment>
`;
